Show func and method signature in quick doc (fixes #1234)

dlsniper opened the pull request:
This works only if the documentation is present at the moment.

ignatov commented on src/com/goide/GoDocumentationProvider.java:
```diff
@@ -86,8 +86,8 @@ private static String getPackageComment(@Nullable PsiFile file) {
   }
 
   @NotNull
-  private static String getCommentText(@NotNull List<PsiComment> comments) {
```
Why it's called getCommentText and can return a signature?

dlsniper commented:
@ignatov I've updated the PR with proper handling for generating the signature and I've also changed it to retrieve the function signature regardless if other comments are present or not.

dlsniper commented:
And I've fixed the tests as well for it.

ignatov commented on src/com/goide/GoDocumentationProvider.java:
```diff
@@ -102,6 +104,25 @@ else if (type == GoParserDefinition.MULTILINE_COMMENT) {
+      return "func " +
+             ((GoFunctionDeclaration)element).getIdentifier().getText() +
+             ((GoFunctionDeclaration)element).getSignature().getText() +
+             "<br/><br/>";
```
Why two `<br/>`?

dlsniper replied:
I was looking at the screenshot in the comment here: https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/1234#issuecomment-72492954 and trying to give it a distinction in relation to the documentation.
Should I rather bold the the function signature?

ignatov replied:
Please attach a screenshot with current situation.

Sergey Ignatov

On 24 February 2015 at 19:16, Florin Patan <notifications@github.com> wrote:

> In src/com/goide/GoDocumentationProvider.java
> <https://github.com/go-lang-plugin-org/go-lang-idea-plugin/pull/1367#discussion_r25265059>
> :
>
> > @@ -102,6 +104,25 @@ else if (type == GoParserDefinition.MULTILINE_COMMENT) {
> >          }
> >          return text;
> >        }
> > -    }), "<br/>") + "</pre>";
> > +    }), "<br/>");
> > +  }
> > +
> > +  private static String getSignature(PsiElement element) {
> > +    if (element instanceof GoFunctionDeclaration) {
> > +      return "func " +
> > +             ((GoFunctionDeclaration)element).getIdentifier().getText() +
> > +             ((GoFunctionDeclaration)element).getSignature().getText() +
> > +             "<br/><br/>";
>
> I was looking at the screenshot in the comment here: #1234 (comment)
> <https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/1234#issuecomment-72492954>
> and trying to give it a distinction in relation to the documentation.
> Should I rather bold the the function signature?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/go-lang-plugin-org/go-lang-idea-plugin/pull/1367/files#r25265059>
> .
>

dlsniper replied:
Please see below:

![snapshot6](https://cloud.githubusercontent.com/assets/607868/6353441/8725b96c-bc49-11e4-9904-35022be52469.png)

linniksa commented:
![123](https://cloud.githubusercontent.com/assets/424343/6353323/b6647636-bc59-11e4-9fb1-0521a7763a11.png)
Maybe add signature to &lt;b> tag and format doc and signature by 2 &lt;pre> tags istead of br:
![123](https://cloud.githubusercontent.com/assets/424343/6353430/3bf41a36-bc5a-11e4-915e-1b9596b6e88b.png)

linniksa commented:
Also want to point out that the current implementation of parsing comments not the same with godoc in many points. I think need create issue do work on rewrite.
but so far I have too little time for active participation.

ignatov commented:
@dlsniper @linniksa The different start offset looks unpleasant.

dlsniper commented:
@linniksa I think this is true for documentation that is provided via ```doc.go``` inside a package.
At least that's where I've seen issues with the documentation. Extracting a function doc from a ```doc.go``` would be very hard I think as it has no structure. The only thing that I can think for that would be to provide a tool window with all the package doc displayed in it.

linniksa commented:
@ignatov simple way is add space to signature at this time
godoc strip same whitespaces prefix form all lines, and if line contains more whitespace characters wrap it in to block (pre)

dlsniper commented:
I've changed it to something like this:

![snapshot7](https://cloud.githubusercontent.com/assets/607868/6353829/e471de46-bc4b-11e4-84e9-a30df9a13f71.png)

ignatov commented:
What're thinking about multiline declarations?

linniksa commented:
at this time i can make comment with first tab char, or 2 spaces or with no spaces
and without complete rewrite of comments processing (for example we do not need add <br> of each line and place all comment in pre tag by default) we cannont get "right" result in that cases

linniksa commented:
@dlsniper 
a this time this nit show signature for interface method

dlsniper commented:
@ignatov we now have support for multi-line defined functions (maybe not in a nice way, http://9gag.com/gag/aLPD0Xg?ref=fbp                                     ) 
![snapshot8](https://cloud.githubusercontent.com/assets/607868/6357489/2e6d1060-bc65-11e4-9bbe-107d8c672733.png)

ignatov commented on src/com/goide/GoDocumentationProvider.java:
```diff
@@ -44,7 +44,14 @@ public String generateDoc(PsiElement element, PsiElement originalElement) {
+        replaceAll("\n", "").
+        replaceAll("\\s+", " ").
+        replaceAll("\\(\\s", "(").
+        replaceAll(",\\s\\)", ")");
```
It's not a nice way to create a one-liners. Please use PSI for this. See GoParameterInfoHandler.

dlsniper replied:
Ah, you are right, I've forgot to check where we do that. Thanks.

dlsniper commented:
I've updated the PR to fix the previous comments and added support for interface signatures as well (good find @linniksa, thank you ;) )

ignatov commented on src/com/goide/GoDocumentationProvider.java:
```diff
@@ -102,6 +104,64 @@ else if (type == GoParserDefinition.MULTILINE_COMMENT) {
+    return result + "</b><br/>";
+  }
+  
+  private static String getParametersAsString(GoParameters parameters) {
```
Can we reuse some code from GoParameterInfoHandler?

dlsniper replied:
The whole building part should be reusable, that's why I extracted it in a separate method.
The issue is that GoParameterInfoHandler::updatePresentation() needs some rework in order to properly signal where the current parameter is, which unfortunately is tightly coupled with building the parameter representation.
I'll look into it.

zolotov replied:
As an example you can use https://github.com/JetBrains/intellij-community/blob/master/java/java-impl/src/com/intellij/codeInsight/hint/api/impls/AnnotationParameterInfoHandler.java, take a look at static `updateUIText` method.

ignatov commented on src/com/goide/GoDocumentationProvider.java:
```diff
@@ -102,6 +104,64 @@ else if (type == GoParserDefinition.MULTILINE_COMMENT) {
+  }
+
+  @NotNull
+  private static String getSignature(PsiElement element) {
```
Please annotate your code via `Infer nullity` action.

linniksa commented:
@dlsniper 
result with unnamed parameters not shown:
```go
func Foo() string {

}

func Bar() (string, error) {

}
```

dlsniper commented:
@linniksa thanks. I'll work on the PR to update it with further tests to cover all scenarios and hopefully it will be done in time for the next alpha release.

ignatov commented on src/com/goide/GoDocumentationProvider.java:
```diff
@@ -102,6 +104,64 @@ else if (type == GoParserDefinition.MULTILINE_COMMENT) {
+      return "";
+    }
+
+    String result = " <b>func " + (identifier != null ? identifier.getText() : "") + "(";
```
Looks like you've forgot about methods.

dlsniper replied:
I've took the decision to not display the receiver of the function as I don't think it's relevant for the user to know that information. I've tested in practice both with and without it before committing it like this.

ignatov commented on src/com/goide/GoDocumentationProvider.java:
```diff
@@ -102,6 +104,64 @@ else if (type == GoParserDefinition.MULTILINE_COMMENT) {
+
+    return result + "</b><br/>";
+  }
+  
```
@NotNull?

dlsniper replied:
I've used the ```Infer nullity``` action and that's how it annotated the code.

dlsniper commented:
I'd need some help with the tests as I can't figure out what I've done wrong compared with the other examples. Thank you.

zolotov commented:
Obviously you have no reference at the caret. And it shouldn't be there actually, it's function declaration.

dlsniper commented:
I've fixed the tests and the rest of the feedback on this one. Unless there's anything that I'm missing, this can be merged as well. Thanks.

zolotov commented:
I'm going to merge this. But there are three pieces of code that do similar things:
1. GoDocumentationProvider#getParametersAsString
2. GoParameterInfoHandler#updatePresentation
3. GoCompletionUtil#createFunctionOrMethodLookupElement – calculating paramTest

Could you handle it please?

dlsniper commented:
Sure, I'll create a separate PR to unify those, thanks

zolotov commented on src/com/goide/GoDocumentationProvider.java:
```diff
@@ -102,6 +104,78 @@ else if (type == GoParserDefinition.MULTILINE_COMMENT) {
+    }
+
+    StringBuilder builder = new StringBuilder();
+    for (int i = 0; i < paramPresentations.size(); ++i) {
```
StringUtil#join?
